### ISSUE TITLE
Getting infinite redirects for "pages" sub-folders?

### ISSUE THREAD

**Issue by michael-newsrx** [gitbucket/gitbucket-pages-plugin#30]
I'm trying to use the "pages" plugin to display mkdocs generated documentation.
However, whenever a sub-folder is accessed that has an index.html in it I am seeing a redirect instead of it loading the index.html document. If I specify the index.html document after the final slash it will load fine.

This is behind an Apache AJP proxy.

GITBUCKET_VERSION | 4.35.3

```
GET /gitbucket/My-Organization/Documentation/pages/print_page/ HTTP/1.1" 302
... over and over ...
```

Backend (tomcat) access log:

```
50.241.2.217 - - [14/Jan/2021:09:31:47 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/js/print-site.js HTTP/1.1" 200 3327
50.241.2.217 - - [14/Jan/2021:09:31:47 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/assets/javascripts/worker/search.8c7e0a7e.min.js HTTP/1.1" 200 35881
50.241.2.217 - - [14/Jan/2021:09:31:47 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/search/search_index.json HTTP/1.1" 200 2062
50.241.2.217 - - [14/Jan/2021:09:31:47 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/assets/images/favicon.png HTTP/1.1" 200 1870
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:09 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
50.241.2.217 - - [14/Jan/2021:09:32:10 -0500] "GET /gitbucket/NewSRX-Tech-LLC/Documentation/pages/print_page/ HTTP/1.1" 302 -
```

AJP config:

```
ProxyPreserveHost On
ProxyPass /gitbucket/ ajp://tomcat-0002:8009/gitbucket/ connectionTimeout=30 timeout=600
ProxyPass /gitbucket ajp://tomcat-0002:8009/gitbucket connectionTimeout=30 timeout=600
```

**Comment by yaroot** [comment]
I'm not familiar with both Tomcat and Apache 2, but it seems like a trailing slash problem. Maybe try http on tomcat to see if it works without apache 2.

**Comment by michael-newsrx** [comment]
Why is Tomcat initiating a redirect for a path with a trailing slash instead of serving up the contents of index.html?

Hrm...

My attempts at using a tunnel to directly connect to the Tomcat instance results in the gitbucket redirecting me to the public address as configured in the software for login.

**Comment by michael-newsrx** [comment]
### Workaround

I've come up with a hackish work around.

My httpd.conf file was modified as follows:

```diff
+       RewriteEngine on
+       RewriteRule ^/gitbucket/(.*)/(.*)/pages/(.*)/$ ajp://tomcat:8009/gitbucket/$1/$2/pages/$3/index.html [P]
        ProxyPass /gitbucket/ ajp://tomcat:8009/gitbucket/ connectionTimeout=30 timeout=600
```

**Comment by yaroot** [comment]
The plugin redirects `/foo` to `/foo/` for all the relative url in to work, my guess is that the program does not get the trailing slash. I guess the application did not actually get the trailing slash.

**Comment by michael-newsrx** [comment]
> The plugin redirects `/foo` to `/foo/` for all the relative urls to work, my guess would be that the program didn't get the trailing slash.

The tomcat log shows the requests that are getting redirected ending with a "/".

**Comment by yaroot** [comment]
The problem is in route matching, we'll address them

**Comment by yaroot** [comment]
Fixed by #33
